README.md

# StackSplit                            [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5802051.svg)](https://doi.org/10.5281/zenodo.5802051) [![View michaelgrund/stacksplit on File Exchange](https://www.mathworks.com/matlabcentral/images/matlab-file-exchange.svg)](https://de.mathworks.com/matlabcentral/fileexchange/62402-michaelgrund-stacksplit)          

### A plugin for multi-event shear wave splitting analyses in SplitLab     

StackSplit is a plugin for the MATLAB toolbox [SplitLab](http://splitting.gm.univ-montp2.fr/) ([**_Wüstefeld et al., 2008_**](https://www.sciencedirect.com/science/article/pii/S0098300407001859)) which allows to apply multi-event techniques for shear wave splitting measurements (SWS) directly from within the main program. 

For details regarding the installation and usage, see the [UserGuide](https://github.com/michaelgrund/stacksplit/blob/main/StackSplit/Doc/StackSplit_userguide.md). 

Citation
--------

If you make use of StackSplit in your work please acknowledge my paper in which the program is described:

- **_Grund, M. (2017)_**, StackSplit - a plugin for multi-event shear wave splitting analyses in SplitLab, *Computers & Geosciences*, 105, 43-50, https://doi.org/10.1016/j.cageo.2017.04.015.

Optionally, you can also cite the [Zenodo DOI](https://zenodo.org/record/5802051#) given above which is referring the latest version of this GitHub repository.

Which stacking methods are available?
-------------------------------------

StackSplit grants easy access to four stacking schemes with which single SWS measurements made with SplitLab can be processed:

1. **WS**: stacking of error surfaces, normalized on minimum/maximum (depending on input) of each single surface ([**_Wolfe & Silver, 1998_**](https://doi.org/10.1029/97JB02023))

2. **RH**: modified WS method with weight depending on SNR of each measurement and normalization regarding the available backazimuth directions ([**_Restivo & Helffrich, 1999_**](https://doi.org/10.1046/j.1365-246x.1999.00845.x))

3. **no weight**: stacking of error surfaces without weighting following PhD thesis of [**_Wüstefeld (2007)_**](http://splitting.gm.univ-montp2.fr/)

4. **SIMW**: simultaneous inversion of multiple waveforms in time domain ([**_Roy et al., 2017_**](https://doi.org/10.1093/gji/ggw470))

![fig4github](https://user-images.githubusercontent.com/23025878/56716351-6d3d2a80-673a-11e9-8b34-2191c119d780.png)

Compatibility with SplitLab and MATLAB versions
-----------------------------------------------

|StackSplit|SplitLab|MATLAB|
|---|---|---|
|dev ([main branch](https://github.com/michaelgrund/stacksplit))|[1.2.1](https://robporritt.wordpress.com/software/), [1.0.5](http://splitting.gm.univ-montp2.fr/) (not tested)|>= [2020a](https://mathworks.com/help/releases/R2020a/index.html) (< 2020a might work, but not tested yet)|
|[v3.0](https://github.com/michaelgrund/stacksplit/releases/tag/v3.0) (latest release)|[1.2.1](https://robporritt.wordpress.com/software/), [1.0.5](http://splitting.gm.univ-montp2.fr/) (not tested)|>= [2020a](https://mathworks.com/help/releases/R2020a/index.html) (< 2020a might work, but not tested yet)|
|[v2.0](https://github.com/michaelgrund/stacksplit/releases/tag/v2.0)|[1.2.1](https://robporritt.wordpress.com/software/), [1.0.5](http://splitting.gm.univ-montp2.fr/)|>= [2014b](https://mathworks.com/company/newsroom/mathworks-introduces-new-features-in-matlab-and-simulink.html) (tested up to and including [2018b](https://mathworks.com/help/releases/R2018b/index.html))|
|[v1.0](https://github.com/michaelgrund/stacksplit/releases/tag/v1.0)|[1.2.1](https://robporritt.wordpress.com/software/), [1.0.5](http://splitting.gm.univ-montp2.fr/)|<= [2014a](https://mathworks.com/company/newsroom/mathworks-announces-release-2014a-of-the-matlab-and-simulink-product-families.html)|

For details regarding the different StackSplit versions see the [Changelog](https://github.com/michaelgrund/stacksplit/blob/main/changelog.md).

Contributing
------------

Did you find a bug or have suggestions for improvements? Simply open a new [issue](https://github.com/michaelgrund/stacksplit/issues) or [pull request](https://github.com/michaelgrund/stacksplit/pulls) here on GitHub.

Related topics
--------------

- The most recent SplitLab version can be found here (not compatible with StackSplit yet): https://github.com/IPGP/splitlab

- Shear wave splitting analysis in Python (based on SplitLab): https://github.com/paudetseis/SplitPy

- Shear wave splitting analysis in Julia: https://github.com/anowacki/SeisSplit.jl

References
----------

- **_Restivo, A. & Helffrich, G. (1999)_**, Teleseismic shear wave splitting measurements in noisy environments, Geophysical Journal International 137, 821-830, https://doi.org/10.1046/j.1365-246x.1999.00845.x.
- **_Roy, C., Winter, A., Ritter, J. R. R., Schweitzer, J. (2017)_**, On the improvement of SKS splitting measurements by the simultaneous inversion of multiple waveforms (SIMW), Geophysical Journal International, 208, 1508–1523, https://doi.org/10.1093/gji/ggw470.
- **_Wolfe, C. J. & Silver, P. G. (1998)_**, Seismic anisotropy of oceanic upper mantle: Shear wave splitting methodologies and observations, Journal of Geophysical Research 103(B1), 749-771, https://doi.org/10.1029/97JB02023.
- **_Wüstefeld, A. (2007)_**, Methods and applications of shear wave splitting: The East European Craton. Ph.D. thesis, Univ. de Montpellier, France, http://splitting.gm.univ-montp2.fr/.
- **_Wüstefeld, A., Bokelmann, G., Zaroli, C., Barruol, G. (2008)_**, SplitLab: A shear-wave splitting environment in Matlab, Computers & Geosciences 34, 515–528, https://doi.org/10.1016/j.cageo.2007.08.002.


changelog.md
# Changelog

## Release [v3.0](https://github.com/michaelgrund/stacksplit/releases/tag/v3.0) (2021-12-23) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5802051.svg)](https://doi.org/10.5281/zenodo.5802051)

* adjusted and updated several StackSplit functions to work on newer MATLAB versions (>=2020a)
* removed or replaced deprecated built-in MATLAB functions 
* fixed start time extraction by SplitLab (for details see [**Fröhlich et al., 2022**](https://www.annalsofgeophysics.eu/index.php/annals/article/view/8781) and the corresponding [repo](https://github.com/yvonnefroehlich/SplitLab-TemporalAlignment))
* added warning message box if non-nulls and nulls are selected together for stacking (which is not really reasonable)
* added warning message box if current screen resolution does not allow to display StackSplit's main panel properly:
  * Solution for **Windows 10**: under *Settings* => *System* => *Display* => *Scale and Layout* => *Change the size of text, apps, and other items*
  the selection sometimes is set to a value different from 100% (e.g. 150%)
  which effectively reduces your screen size in pixels. Set it to 125% or
  better 100% and check again, mostly then the panel fits on the screen.
  
**Contributors**: [Michael Grund](https://github.com/michaelgrund), [Yvonne Fröhlich](https://github.com/yvonnefroehlich)

## Release [v2.0](https://github.com/michaelgrund/stacksplit/releases/tag/v2.0) (2019-06-28) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7118716.svg)](https://doi.org/10.5281/zenodo.7118716)

* adjusted several functions to work also on MATLAB 2018a and b
* now results are additionally saved in a text-file which can be used directly in GMT (Generic Mapping Tools)

**Contributors**: [Michael Grund](https://github.com/michaelgrund)

## Release [v1.0](https://github.com/michaelgrund/stacksplit/releases/tag/v1.0) (2017-04-04) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.464385.svg)](https://doi.org/10.5281/zenodo.464385)

StackSplit is now available for download.

**Contributors**: [Michael Grund](https://github.com/michaelgrund)


StackSplit/SS_layout.m
function h=SS_layout(push1_callback,push2_callback,push3_callback,push4_callback,list_callback,...
    check1_callback,check2_callback,check3_callback,check4_callback,pop1_callback,push5_callback,...
    pop2_callback,pop3_callback,pop4_callback,pop5_callback)
%==========================================================================
%##########################################################################
%#                                                                        #
%#  This function is part of StackSplit - a plugin for multi-event shear  #
%#  wave splitting analyses in SplitLab                                   #
%#                                                                        #
%##########################################################################
%==========================================================================
% FILE DESCRIPTION
%
% generate and setup layout for StackSplit plugin 
%
%==========================================================================
% LICENSE
%
% Copyright (C) 2016  Michael Grund, Karlsruhe Institute of Technology (KIT), 
% GitHub: https://github.com/michaelgrund
% 
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <http://www.gnu.org/licenses/>.
% 
% TERMS OF USE
%
% StackSplit is provided "as is" and without any warranty. The author cannot be
% held responsible for anything that happens to you or your equipment. Use it
% at your own risk.
%==========================================================================

%==================================================================================================================================
%================================================================================================================================== 

global config

shift_right=45;

% check if a SS figure window is already open, otherwise create one
h.fig=findobj('type','figure','name',['StackSplit ' config.SS_version]);

if isempty(h.fig)
    h.fig=figure('units','pixels','name',['StackSplit ' config.SS_version], 'numbertitle','off','menubar','none');
    set(h.fig, 'position', [150, 21, 1450, 600])
else
    close(h.fig)
    %.........................................
    SS_stacksplit_start;
    %.........................................
end

set(gcf, 'Resize', 'off');

%==================================================================================================================================  
%==================================================================================================================================                         
% Panel layout

h.panel(1) = uipanel('Units','pixel','Title','Selection window',...
    'FontSize',10,'Position',[14 22 902 568], 'BackgroundColor', [224   223   227]/255);

h.panel(2) = uipanel('Units','pixel','Title','Minimum energy surface',...
    'FontSize',10,'Position',[934 22 502 568], 'BackgroundColor', [224   223   227]/255);

h.panel(3) = uipanel('Parent',h.panel(2),'Units','pixel',...
    'FontSize',10,'Position',[158.5,15,182,49], 'BackgroundColor', 'w');

h.panel(4)= uipanel('Parent',h.panel(1),'Units','pixel','title','Surface stack',...
    'FontSize',10,'Position',[580+shift_right 7 119 170], 'BackgroundColor', [224   223   227]/255,'TitlePosition','centertop');

h.panel(5)= uipanel('Parent',h.panel(1),'Units','pixel','title','Joint inversion',...
    'FontSize',10,'Position',[715+shift_right 76 119 101], 'BackgroundColor', [224   223   227]/255,'TitlePosition','centertop');

h.panel(6) = uipanel('Units','pixel','Title','Waveforms',...
    'FontSize',10,'Position',[934 22 502 568], 'BackgroundColor', [224   223   227]/255);

h.panel(7) = uipanel('Parent',h.panel(1),'Units','pixel',... % panel "around" listbox
    'FontSize',10,'Position',[28,68,577,419], 'BackgroundColor', [224   223   227]/255);

h.panel(8) = uipanel('Parent',h.panel(1),'Units','pixel','Title','Limits',... 
    'FontSize',10,'Position',[600+shift_right,187,214,102], 'BackgroundColor', [224   223   227]/255,'TitlePosition','centertop');

% event info header above listbox
h.ev_info = uicontrol('style','text','horizontalalignment','left');
set(h.ev_info,'String',['                  t0time               JD      BAZ     dist     inipol     inc    SNR         filter          pha'])
set(h.ev_info,'Position',[45,493,572,15],'fontsize',10)
set(h.ev_info,'BackgroundColor',[0.7 0.7 0.7]);

% staname beside worldmap
h.staname_info = uicontrol('style','text','horizontalalignment','left');
set(h.staname_info,'String',config.stnname)
set(h.staname_info,'Position',[790+shift_right,550,68,22],'fontsize',15)

set(h.staname_info,'BackgroundColor',[224   223   227]/255);  

%==================================================================================================================================  
%==================================================================================================================================                         
% Panel << Selection window >> control buttons

h.push(1)=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','pushbutton',...
                             'BackgroundColor','w',...
                             'Position',[85,15,120,35],...
                             'String', 'Stack',...
                             'Callback',push1_callback);

h.push(2)=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','pushbutton',...
                             'BackgroundColor','w',...
                             'Position',[252,15,120,35],...
                             'String', 'Clear',...  
                             'Callback',push2_callback);

h.push(3)=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','pushbutton',...
                             'BackgroundColor','w',...
                             'Position',[419,15,120,35],...
                             'String', 'Save',...  
                             'Callback',push3_callback);

h.push(4)=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','pushbutton',...
                             'BackgroundColor','w',...
                             'Position',[737.5+shift_right,19,72,21],...
                             'String', 'Exit',...  
                             'Callback',push4_callback);
                         
h.push(5)=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','pushbutton',...
                             'BackgroundColor','w',...
                             'Position',[102.5,15,420,35],...
                             'String', 'Inversion',...
                             'Callback',push5_callback);
    
%==================================================================================================================================  
%==================================================================================================================================                         
% Panel << Selection window >> check boxes

h.check(1)=uicontrol('Parent',h.panel(4),'Units','pixel',...
                             'Style','radiobutton',...
                             'BackgroundColor',[224   223   227]/255,...
                             'Position',[13 126 96 17],...
                             'String', 'no weight',...  
                             'Callback',check1_callback);

h.check(2)=uicontrol('Parent',h.panel(4),'Units','pixel',...
                             'Style','radiobutton',...
                             'BackgroundColor',[224   223   227]/255,...
                             'Position',[13 94 91 17],...
                             'String', 'WS (1998)',...  
                             'Callback',check2_callback);
                         
h.check(3)=uicontrol('Parent',h.panel(4),'Units','pixel',...
                             'Style','radiobutton',...
                             'BackgroundColor',[224   223   227]/255,...
                             'Position',[13 59 86 17],...
                             'String', 'RH (1999)',...  
                             'Callback',check3_callback);  
                         
h.check(4)=uicontrol('Parent',h.panel(5),'Units','pixel',...
                             'Style','radiobutton',...
                             'BackgroundColor',[224   223   227]/255,...
                             'Position',[13 58 70 17],...
                             'String', 'SIMW',...  
                             'Callback',check4_callback);   
                                                
%==================================================================================================================================  
%==================================================================================================================================                         
% Panel << Selection window >> popup wdw for SIMW taper selection      

h.pop(1)=uicontrol('Parent',h.panel(5),'Units','pixel',...
                             'Style','popup',...
                             'BackgroundColor','w',...
                             'Position',[15 15 70 17],...
                             'String', 'none|10|20|30|40|50|60|70|80|90|100',...
                             'Callback',pop1_callback); 

h.taptext=uicontrol('Parent',h.panel(5),'Style','text',...
        'BackgroundColor',[224   223   227]/255,...
        'Position',[8 32 70 17],...
        'String','% taper');
       
%==================================================================================================================================  
%==================================================================================================================================                         
% Panel << Selection window >> popup wdw for selection between Emap and EVmap (lambda2map)    

h.pop(4)=uicontrol('Parent',h.panel(4),'Units','pixel',...
                             'Style','popup',...
                             'BackgroundColor','w',...
                             'Position',[15 15 70 17],...
                             'String', 'Esurf|EVsurf',...
                             'Callback',pop4_callback); 

h.inputtext=uicontrol('Parent',h.panel(4),'Style','text',...
        'BackgroundColor',[224   223   227]/255,...
        'Position',[8 33 70 17],...
        'String','Surf input');    
                       
%==================================================================================================================================  
%==================================================================================================================================                         
% Panel << Selection window >> listbox
                   
h.list=uicontrol('Parent',h.panel(1),'Units','pixel',...
                             'Style','listbox',...
                             'BackgroundColor','w',...
                             'Position',[30,70,573,400],...
                             'Callback',list_callback);                           

%==================================================================================================================================  
%==================================================================================================================================                                                
% Panel << Error map >> axes                      
                                                
h.axEmap=axes('Parent',h.panel(2),'Units','pixel',...
                             'Position',[56,111,385,411]);

%==================================================================================================================================  
%================================================================================================================================== 
% Panel << Waveforms >> axes               
                                                
h.axWF=axes('Parent',h.panel(6),'Units','pixel',...
                             'Position',[60,300,400,225]);


h.axWF2=axes('Parent',h.panel(6),'Units','pixel',...
                              'Position',[60,50,400,225]);                        
                         
%==================================================================================================================================  
%==================================================================================================================================                       
% Panel << Selection window, Limits >> popup wdw for max BAZ, max dist and max pol   
   
h.pop(2)=uicontrol('Parent',h.panel(8),'Units','pixel',...
                             'Style','popup',...
                             'BackgroundColor','w',...
                             'Position',[123 65 70 17],...
                             'String', 'none|1|2|3|4|5|6|7|8|9|10',...
                             'Callback',pop2_callback); 

h.maxbaztext1=uicontrol('Parent',h.panel(8),'Style','text',...
        'BackgroundColor',[224   223   227]/255,...
        'Position',[11 61 100 17],...
        'String',['max ' char(hex2dec('0394')) 'BAZ in °']);       
    
%............................................................................    

h.pop(3)=uicontrol('Parent',h.panel(8),'Units','pixel',...
                             'Style','popup',...
                             'BackgroundColor','w',...
                             'Position',[123 40 70 17],...
                             'String', 'none|1|2|3|4|5|6|7|8|9|10',...
                             'Callback',pop3_callback); 

h.maxdisttext1=uicontrol('Parent',h.panel(8),'Style','text',...
        'BackgroundColor',[224   223   227]/255,...
        'Position',[8 35 100 17],...
        'String',['max ' char(hex2dec('0394')) 'dist in °']);  
       
%............................................................................    
% limit for initial polarization    
    
h.pop(5)=uicontrol('Parent',h.panel(8),'Units','pixel',...
                             'Style','popup',...
                             'BackgroundColor','w',...
                             'Position',[123 15 70 17],...
                             'String', 'none|1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20',...
                             'Callback',pop5_callback); 

h.maxpoltext1=uicontrol('Parent',h.panel(8),'Style','text',...
        'BackgroundColor',[224   223   227]/255,...
        'Position',[13 9 100 17],...
        'String',['max ' char(hex2dec('0394')) 'inipol in °']);   
    
end

%==================================================================================================================================  
%==================================================================================================================================  
% EOF